### DISCUSSION TITLE
How to detect if function f(x) blows up to positive infinity (or negative infinity) over a finite-sized domain of x?

### DISCUSSION THREAD

**Discussion by unary-code** [MilesCranmer/PySR#366] [Q&A] [answered]
I am trying to use the PySR regressor model to find a function.
Now, I am only giving the regressor a domain from x=0 to x=1, because I figure that the model would be able to better predict a function that only exists over a finite-sized domain than a function that exists over the infinite-sized domain (-infinity, +infinity).

I want to tell the model to not accept (to not return) a function which blows up to positive infinity or negative infinity.

I tried one approach (but it ended up not working, because the PySR model still found a final function that does blow up to infinity, so this approach has a possibility of false negatives, where the approach fails to detect a function that does in fact blow up to infinity). This approach is basically: In my custom objective function if the value of the value of the guessed function
1. on any one of the dataset points, OR
2. on any one of the x values 0/200, 1/200, 2/200, up to 200/200
is greater than 10000 or less than 0, then I return positive infinity in Julia to signal to the model to reject this function.

But obviously this approach is flawed because it's very likely that the model will just change the values of certain parameters so slightly so that the function blows up in reality over the real-values from 0 to 1 but does NOT blow up over any one of the dataset points and does NOT blow up on any one of the x values 0/200, 1/200, 2/200, up to 200/200.

For example, the given dataset is 100 data points. Each data point is 1-dimensional, has a real-value from the uniform distribution over x=[0,1). So that means the value of all my datapoints are in between 0 (inclusive) and 1 (exclusive).

Note that my objective function first squares the function represented by the tree "tree", then (badly) approximates its integral from x=0 to x=1 which is its approximation of the normalization constant, then divides by that normalization constant.
```
objective = """
function eval_loss(tree, dataset::Dataset{T,L}, options)::L where {T,L}
    MAX_THRESHOLD_ALLOWED = 10000

    prediction, flag = eval_tree_array(tree, dataset.X, options)
    !flag && return L(Inf)
    
    num_points = length(prediction)
    
    for i in 1:num_points
        cur_expr_val = prediction[i]
        if (cur_expr_val > MAX_THRESHOLD_ALLOWED)
            return Inf
        end
        if (cur_expr_val < 0)
            return Inf
        end
    end
    
    num_rectangles = 200
    
    # You make sure that this object is of type Matrix, not Vector and not Array.
    evenly_spaced_numbers = reshape([LinRange(0, 1, num_rectangles + 1);], 1, num_rectangles + 1)
    evenly_spaced_numbers = Float32.(evenly_spaced_numbers)
    
    prediction_on_evenly_spaced_numbers, flag_on_evenly_spaced_numbers = eval_tree_array(tree, evenly_spaced_numbers, options)
    
    

    norm_constant = 0
    prev_expr_val = -1
    for i in 0:num_rectangles
        cur_expr_val = prediction_on_evenly_spaced_numbers[i+1]
        if (cur_expr_val > MAX_THRESHOLD_ALLOWED)
            return Inf
        end
        if (cur_expr_val < 0)
            return Inf
        end
        
        cur_expr_val = cur_expr_val * cur_expr_val
        if (i > 0)
            cur_trapezoid_area = (cur_expr_val + prev_expr_val) / (2.0 * num_rectangles)
            norm_constant += cur_trapezoid_area
        end
        
        prev_expr_val = cur_expr_val
    end
    
    prediction = (prediction .* prediction)
    
    actual_probs = (prediction) / norm_constant
    
    # println("HELLO WORLD")
    
    # length(actual_probs) equals length(prediction)
    return exp(-1 * sum(log.(actual_probs)) / (length(prediction)))
end
"""
model = PySRRegressor(
    niterations=40,  # < Increase me for better results
    binary_operators=["+", "*", "-", "/"],
    unary_operators=[
         "exp",
     ],
    # ^ Define operator for SymPy as well
    full_objective=objective
    # ^ Custom loss function (julia syntax)
)
```

This resulted in the model ending up finding this function:
"4  >>>>       inf  exp(-0.0075005013 / (0.45609152 - x0))  0.000000   6"
The score was "inf", the loss was exactly 0.

The orange line is the model's guessed function: exp(-0.0075005013/(0.45609152-x))
The red line is the model's guessed function but squared and then divided by the code's approximation of the normalization constant 1.2598633:
The green line is the true probability function (the uniform function over x=[0, 1)).

My thoughts for ways to get around this: Possibly I could use calculus? I know about Newton's method to detect x-values of roots. If I take the reciprocal of the model's guessed function, then I implement Newton's method, perhaps in my objective function I could be able to detect more functions that blow up and reject them by returning positive infinity in my objective function.

<img width="541" alt="Screen Shot 2023-07-02 at 6 35 16 PM" src="https://github.com/MilesCranmer/PySR/assets/60202695/7d353449-d03b-4ea6-80cb-5b607fb561f3">

**Comment by MilesCranmer** [comment, marked as answer]
I guess the issue is that it is finding ways to exploit inaccuracies in your integration by putting in blow-ups outside of the points you sample? One option is you could use an adaptive integrator like QuadGK.jl inside the objective function: https://github.com/JuliaMath/QuadGK.jl. See https://astroautomata.com/PySR/examples/#7-julia-packages-and-types for using external packages in the objective. You would probably want to limit the number of steps it can take, and simply return `L(Inf)` if the integration does not converge.

**Reply by unary-code** [reply]
Thanks for the fast and helpful response! I will try out QuadGK now and I'll let you know if that works to properly detect functions that blow up to infinity (at least over the domain x=[0, 1) ).
